libkbfs/prefetcher_test.go

```diff
@@ -1066,9 +1077,401 @@ func TestSyncBlockCacheWithPrefetcher(t *testing.T) {
+	encA, serverHalfA := setupRealBlockForDiskCache(t, aPtr, a, dbcConfig)
+	encB, serverHalfB := setupRealBlockForDiskCache(t, bPtr, b, dbcConfig)
+
+	_, _ = bg.setBlockToReturn(rootPtr, root)
+	_, _ = bg.setBlockToReturn(aPtr, a)
+	_, _ = bg.setBlockToReturn(bPtr, b)
+	err := dbc.Put(ctx, kmd.TlfID(), rootPtr.ID, encRoot, serverHalfRoot)
+	require.NoError(t, err)
+	err = dbc.Put(ctx, kmd.TlfID(), aPtr.ID, encA, serverHalfA)
+	require.NoError(t, err)
+	err = dbc.Put(ctx, kmd.TlfID(), bPtr.ID, encB, serverHalfB)
+	require.NoError(t, err)
+
+	t.Log("Fetch dir root.")
+	var block Block = &DirBlock{}
+	ch := q.Request(context.Background(), defaultOnDemandRequestPriority, kmd,
+		rootPtr, block, TransientEntry)
+	notifySyncCh(t, prefetchSyncCh)
+	err = <-ch
+	require.NoError(t, err)
+	// Notify sync channel for the child blocks `a` and `b`.
+	notifySyncCh(t, prefetchSyncCh)
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Cancel the prefetch for block 'a'.")
+	q.Prefetcher().CancelPrefetch(aPtr.ID)
+	// Notify sync channel for the cancelation.
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Set the metadata of the root block in the disk cache to " +
+		"NoPrefetch, simulating an eviction and a BlockServer.Get.")
+	err = dbc.UpdateMetadata(ctx, rootPtr.ID, NoPrefetch)
+	require.NoError(t, err)
+
+	t.Log("Evict the root block from the block cache.")
+	err = bcache.DeleteTransient(rootPtr, kmd.TlfID())
+	require.NoError(t, err)
+
+	t.Log("Fetch dir root again.")
+	block = &DirBlock{}
+	ch = q.Request(context.Background(), defaultOnDemandRequestPriority, kmd,
+		rootPtr, block, TransientEntry)
+	notifySyncCh(t, prefetchSyncCh)
+	err = <-ch
+	require.NoError(t, err)
+	// Notify sync channel for only child block `a`, since `b` wasn't newly
+	// triggered.
+	notifySyncCh(t, prefetchSyncCh)
+
+	t.Log("Fetch child block \"a\" on demand.")
+	block = &FileBlock{}
+	ch = q.Request(context.Background(), defaultOnDemandRequestPriority, kmd,
+		aPtr, block, TransientEntry)
+	// Notify sync channel for the child block `a`.
+	notifySyncCh(t, prefetchSyncCh)
+	err = <-ch
+	require.NoError(t, err)
+
+	t.Log("Fetch child block \"b\" on demand.")
+	block = &FileBlock{}
+	ch = q.Request(context.Background(), defaultOnDemandRequestPriority, kmd,
+		bPtr, block, TransientEntry)
+	// Notify sync channel for the child block `b`.
+	notifySyncCh(t, prefetchSyncCh)
+	err = <-ch
+	require.NoError(t, err)
+
+	testPrefetcherCheckGet(t, config.BlockCache(), rootPtr, root,
+		FinishedPrefetch, TransientEntry)
+	testPrefetcherCheckGet(t, config.BlockCache(), aPtr, a, FinishedPrefetch,
+		TransientEntry)
+	testPrefetcherCheckGet(t, config.BlockCache(), bPtr, b, FinishedPrefetch,
+		TransientEntry)
+
+	// Then we wait for the pending prefetches to complete.
+	waitForPrefetchOrBust(t, q.Prefetcher().Shutdown())
+}
+
+func TestPrefetcherUnsyncedPrefetchRootEvictedCanceled(t *testing.T) {
+	t.Log("Regression test for KBFS-2588: when a prefetched block has " +
```

Comment:
This summary is exactly the same as for the test above (except no "partial"), can you differentiate it somehow?